src/cpp/include/nodegui/QtWidgets/QTableWidget/qtablewidget_wrap.h

```diff
@@ -0,0 +1,60 @@
+#pragma once
+
+#include <napi.h>
+#include <stdlib.h>
+
+#include <QPointer>
+
+#include "Extras/Utils/nutils.h"
+#include "QtWidgets/QTableWidget/ntablewidget.hpp"
+#include "QtWidgets/QWidget/qwidget_macro.h"
+
+class QTableWidgetWrap : public Napi::ObjectWrap<QTableWidgetWrap> {
+ private:
+  QPointer<NTableWidget> instance;
+
+ public:
+  static Napi::Object init(Napi::Env env, Napi::Object exports);
+  QTableWidgetWrap(const Napi::CallbackInfo& info);
+  ~QTableWidgetWrap();
+  NTableWidget* getInternalInstance();
+  // class constructor
+  static Napi::FunctionReference constructor;
+  // wrapped methods
+  Napi::Value selectedRanges(const Napi::CallbackInfo& info);
+  Napi::Value closePersistentEditor(const Napi::CallbackInfo& info);
+  Napi::Value editItem(const Napi::CallbackInfo& info);
+  Napi::Value setCellWidget(const Napi::CallbackInfo& info);
+  Napi::Value setItem(const Napi::CallbackInfo& info);
+  Napi::Value setHorizontalHeaderItem(const Napi::CallbackInfo& info);
+  Napi::Value setHorizontalHeaderLabels(const Napi::CallbackInfo& info);
+  Napi::Value setVerticalHeaderItem(const Napi::CallbackInfo& info);
+  Napi::Value setVerticalHeaderLabels(const Napi::CallbackInfo& info);
+  Napi::Value clear(const Napi::CallbackInfo& info);
+  Napi::Value clearContents(const Napi::CallbackInfo& info);
+  Napi::Value insertColumn(const Napi::CallbackInfo& info);
+  Napi::Value removeColumn(const Napi::CallbackInfo& info);
+  Napi::Value insertRow(const Napi::CallbackInfo& info);
+  Napi::Value removeRow(const Napi::CallbackInfo& info);
+  Napi::Value scrollToItem(const Napi::CallbackInfo& info);
+  // FROM TABLEVIEW
```

Comment:
Makes sense. We can refactor this when we implement tableview. 

Since, the inheritance chain for qtablewidget looks like this.

qtablewidget -> qtableview -> qabstractItemView -> QAbstractScrollArea
We can use the QAbstractScrollArea here already until we implement  qtableview and qabstractItemView.

So if you take a look at QScrollArea and use the QAbstractScrollArea macros instead, we could get all those methods aswell. @slidinghotdog 

By the way this is a lot of good work!

Reply:
Yeah, i saw the inheritance, and most methods are table-oriented and i think they come from QTableView, but some come from QAbstractItemView, but i didn't understand how to use QModelIndex to implement those.
I just commented on that because i copied the methods from QTableView docs page...

Also is there a way to return the JS widget from c++? I saw some methods that return a widget but they were managed on JS side, like `viewport()` on `QAbstractScrollArea.ts`

Reply:
> Yeah, i saw the inheritance, and most methods are table-oriented and i think they come from QTableView, but some come from QAbstractItemView, but i didn't understand how to use QModelIndex to implement those.
> I just commented on that because i copied the methods from QTableView docs page...
> 
Yep, So my point was that we leave QAbstractItemView and QTableView as it is at the moment. We just use the QAbstractScollArea macro instead of QWidget macro on the c++ side. And on the JS side instead of inheriting from nodewidget, we inherit from qabstractscollarea. Nothing else changes.

> Also is there a way to return the JS widget from c++? I saw some methods that return a widget but they were managed on JS side, like viewport() on QAbstractScrollArea.ts

In case of things like QClipboard, QImage etc its pretty easy to return stuff.
But it is slightly tricky in the case of classes that inherit from QObject (which are all widgets).

Which exact method are you talking about since this would vary depending on the context.

Reply:
>We just use the QAbstractScollArea macro instead of QWidget macro on the c++ side. And on the JS side instead of inheriting from nodewidget, we inherit from qabstractscollarea. Nothing else changes.

Oh that makes more sense, i can change that tonight..

>Which exact method are you talking about since this would vary depending on the context.

For methods like these ones [here](https://doc.qt.io/qt-5/qtablewidget.html)

```
 takeHorizontalHeaderItem(int column)
 takeItem(int row, int column)
 takeVerticalHeaderItem(int row)
```
I think i would need to keep a table on ts class with the data (maybe an array of arrays instead of Set) to be able to return them.

Reply:
Nice, QTableWidgetItem is relatively easier to implement since its similar to QClipboard and other objects (it doesnt inherit from QObject). 
You can follow the case of QClipboard:

  https://github.com/nodegui/nodegui/blob/9aa53ea1185fac5978725c7dc9fb7e3ededfa900/src/cpp/lib/QtGui/QApplication/qapplication_wrap.cpp#L95

For this to work correctly you would also need to 
implement Napi::External support in the constructor of QTableWidgetItem 

https://github.com/nodegui/nodegui/blob/9aa53ea1185fac5978725c7dc9fb7e3ededfa900/src/cpp/lib/QtGui/QClipboard/qclipboard_wrap.cpp#L26

and in js also here
https://github.com/nodegui/nodegui/blob/9aa53ea1185fac5978725c7dc9fb7e3ededfa900/src/lib/QtGui/QClipboard.ts#L8

Reply:
So Napi::External is a way to pass pointers or any raw c++ value to the js world

Reply:
Thanks, i'll try that!

Reply:
I looked into it and from what i understood i would need to instantiate a new js class with the pointer, or keep a reference to the existing class to return..
`takeItem` returns a qt class and i dont think i can get the js class from that, so i need to create a new js class and return that.. I'll think about it a bit more, but i won't add it for now.. Anything else needs fixing for this PR?

Reply:
It looks good to me, just needs a rebase :) and we can merge

Reply:
And the `demo.ts`? Do u prefer i leave with the new feature or revert before sending a PR?

Reply:
You can just leave your updated copy. Helps in testing. I ll typically clean it up after sometime